src/cli/mod.rs

```diff
@@ -0,0 +1,216 @@
+    trace!("cli configuration - {:?}", cfg);
+    Ok(cfg)
+}
+
+/// CLI setup, covering all sub-commands and arguments.
+fn cli_setup<'a, 'b>() -> App<'a, 'b> {
+    App::new("Afterburn")
+        .version(crate_version!())
+        .arg(
+            Arg::with_name("provider")
+                .long("provider")
+                .help("The name of the cloud provider")
+                .global(true)
+                .takes_value(true),
+        )
+        .arg(
+            Arg::with_name("cmdline")
+                .long("cmdline")
+                .global(true)
+                .help("Read the cloud provider from the kernel cmdline"),
+        )
+        .subcommand(
+            SubCommand::with_name("multi")
+                .about("Perform multiple tasks in a single call")
+                .arg(
+                    Arg::with_name("legacy-cli")
+                        .long("legacy-cli")
+                        .help("Whether this command was translated from legacy CLI args")
+                        .hidden(true),
+                )
+                .arg(
+                    Arg::with_name("attributes")
+                        .long("attributes")
+                        .help("The file into which the metadata attributes are written")
+                        .takes_value(true),
+                )
+                .arg(
+                    Arg::with_name("check-in")
+                        .long("check-in")
+                        .help("Check-in this instance boot with the cloud provider"),
+                )
+                .arg(
+                    Arg::with_name("hostname")
+                        .long("hostname")
+                        .help("The file into which the hostname should be written")
+                        .takes_value(true),
+                )
+                .arg(
+                    Arg::with_name("network-units")
+                        .long("network-units")
+                        .help("The directory into which network units are written")
+                        .takes_value(true),
+                )
+                .arg(
+                    Arg::with_name("ssh-keys")
+                        .long("ssh-keys")
+                        .help("Update SSH keys for the given user")
+                        .takes_value(true),
+                ),
+        )
+}
+
+/// Translate command-line arguments from legacy mode.
+///
+/// In legacy mode there are not sub-commands, and single-dash (Golang-style)
+/// arguments are allowed too.
+fn translate_legacy_args(cli: impl IntoIterator<Item = String>) -> impl Iterator<Item = String> {
+    // Process the first two arguments and check whether there is a sub-command (normal mode)
+    // or not (legacy mode).
+    let mut argv = cli.into_iter();
+    let argv_0 = argv.next().unwrap_or_else(|| "afterburn".to_string());
+    let argv_1 = argv.next();
+    let legacy_mode = match argv_1 {
+        Some(ref arg) => arg.starts_with('-'),
```

Comment:
Not sure if I follow this logic. The "non-legacy" mode could also fit this, no? E.g. `afterburn --cmdline [subcmd]` ?

Reply:
Ah, you are right. I think I'll need to remove the global status from `--cmdline` and `--provider` and push them down as duplicates in each sub-command. Or do you see a better approach to this?